### PR TITLE
Refactor steno and add `STENO_PROTOCOL = [all|txbolt|geminipr]`

### PR DESCRIPTION
**Pull request by precondition**
## Description

Refactor `quantum/process_keycode/process_steno.*` so that it's easier to read and more firmware space efficient (cut the compile size by ~15% if compiling all protocols and by ~50% if compiling only a single protocol) and give the possibility to users to compile only a single protocol because it's wasteful to compile all available steno protocols even if you plan to use only one — after discussing with some folks on the Plover discord, I came to the conclusion that there are little to no use cases to the ability of switching serial protocols on the fly so using only a single serial protocol is by far the most common use case.

I also fixed a bug wherein the `n_pressed_keys` value given to the `postprocess_steno_user` hook  was off by 1. Speaking of `postprocess_steno_user`, shouldn't we rename this to `post_process_steno_user`? I avoided doing it here as that would be a breaking change but that'd be more consistent with all the other available `post_process_*` hooks.

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by fauxpark** [comment]
This breaks some assumptions the CLI makes about `rules.mk` entries - it would be better to do something like RGB Matrix, ie.
```makefile
STENO_ENABLE = yes|no
STENO_DRIVER = gemini|bolt|etc
```

**Comment by precondition** [comment]
Is `STENO_PROTOCOL = gemini|bolt|all` ok? Because "steno driver" doesn't make much sense or does the CLI expects the _DRIVER suffix?

**Comment by drashna** [comment]
`STENO_PROTOCOL` should be okay.  The main thing is that `STENO_ENABLE` should only be yes/no.

**Comment by precondition** [comment]
The `STENO_PROTOCOL` is indeed cleaner than the `STENO_ENABLE_*` approach :+1:

**Comment by precondition** [comment]
(Last force push was to simply rebase the PR on top of latest develop)

**Review by drashna — APPROVED** [review]

**Review by KarlK90 — APPROVED** [review]
Thanks for the refactor and LGTM!

A additional request, it would be great if this PR could add the data driver equivalents for `STENO_ENABLE` and `STENO_PROTOCOL` to `info_rules.json` along with the json schema validations in `keyboard.jsonschema`. Would you mind adding these?

File: data/mappings/info_rules.json [hunk 1]
```diff
@@ -28,6 +28,8 @@
     "SPLIT_KEYBOARD": {"info_key": "split.enabled", "value_type": "bool"},
     "SPLIT_TRANSPORT": {"info_key": "split.transport.protocol", "to_c": false},
     "WAIT_FOR_USB": {"info_key": "usb.wait_for", "value_type": "bool"},
+    "STENO_ENABLE": {"info_key": "stenography.enabled", "value_type": "bool"},
+    "STENO_PROTOCOL": {"info_key": "stenography.protocol"}
```
Comment (KarlK90) [inline data/mappings/info_rules.json:32]:
```suggestion
    "STENO_PROTOCOL": {"info_key": "stenography.protocol"},
```

Comment (KarlK90) [reply data/mappings/info_rules.json:32]: missing one comma, otherwise LGTM

**Pull request merged**